v-next/hardhat/src/internal/builtin-plugins/solidity/build-system/cache.ts

```diff
@@ -0,0 +1,74 @@
+import path from "node:path";
+
+import {
+  exists,
+  getAccessTime,
+  getAllFilesMatching,
+  getFileSize,
+  readJsonFile,
+  remove,
+  writeJsonFile,
+} from "@ignored/hardhat-vnext-utils/fs";
+
+export class ObjectCache<T> {
+  readonly #path: string;
+  readonly #defaultMaxAgeMs: number;
+  readonly #defaultMaxSize: number;
+
+  constructor(
+    basePath: string,
+    namespace: string,
+    version: string,
+    defaultMaxAgeMs: number = 7 * 24 * 60 * 60 * 1000, // 1 week
+    defaultMaxSize: number = 2 * 1024 * 1024 * 1024, // 2 GB
+  ) {
+    this.#path = path.join(basePath, namespace, version);
+    this.#defaultMaxAgeMs = defaultMaxAgeMs;
+    this.#defaultMaxSize = defaultMaxSize;
+  }
+
+  public async set(key: string, value: T): Promise<void> {
+    const filePath = path.join(this.#path, key);
+    await writeJsonFile(filePath, value);
```

Comment:
Given that these can be pretty large files, I think we should write to a temporary name first, and then rename it to the intended name. That way, if hardhat dies while writing it, we wouldn't corrupt the file.

```
    const filePath = path.join(this.#path, key);
    const tmpFilePath = filePath + ".tmp";
    
    await writeJsonFile(tmpFilePath, value);
    await move(tmpFilePath, filePath)
```